### PR TITLE
Support kubernetes datastore

### LINKED ISSUE #29: Support k8s backend
Unlike the bird+etcd implementation, this implementation only supports etcd datastore backend. 

## Current Behavior
The daemon is hardcoded to use etcd backend.

## Possible Solution
Change the daemon to respect the DATASTORE_BACKEND env variable?

## Context
I'm not sure the status of k8s datastore backend. If highly desirable, I can work on it.

### PR DESCRIPTION
**Pull request by fkakuma**
## Description

This supports kubernetes datastore backend.
We should refactor calico-bgp-daemon generally when libcalico-go
is organized.
This also bumps libcalico-go to v1.6.1 and adds k8s.io/client-go.

Fixes #29

## Todos
- [ ] Tests
- [ ] Documentation

## Release Note

```release-note
Added support for using the Kubernetes API datastore in GoBGP
```

### CODE REVIEW

**Comment by caseydavenport** [comment]
@fkakuma thanks for doing this! Agreed it would be great to refactor once libcalico has support for what is needed here.

CC @toshiiw @gunjan5

**Comment by fkakuma** [comment]
I think that calico-bgp-daemon supports the k8s datastore once at this point in time, as it may take long time that libcalico-go is organized or it may not have any needed functions.
So I put this pull-req. Could you merge this codes?

**Comment by caseydavenport** [comment]
I'd appreciate a look from @toshiiw or @gunjan5 before merging since they know this code better than I do.

**Comment by fkakuma** [comment]
@caseydavenport, thank you for reply. OK, I will wait for the result of their review.

File: ipam.go
Comment (toshiiw) [inline ipam.go:77]: This function is a ipamCache internal function and isn't called from outside.
Why do you need this change?

File: main.go
Comment (toshiiw) [inline main.go]: Shouldn't this be "return nil, err"?

File: k8s.go
Comment (toshiiw) [inline k8s.go:1]: It would be better to note (in the comment or the commit log) some part of this file is from confd/backends/k8s/client.go to ease maintenance.

File: k8s.go
Comment (toshiiw) [inline k8s.go]: time.Sleep would be enough, IMO

File: k8s.go
Comment (toshiiw) [inline k8s.go:172]: The code assumes PodCIDR will never change. Is that ok?

File: k8s.go
Comment (toshiiw) [inline k8s.go:575]: ipPoolData is always of string type. Do we need to use interface here?

File: k8s.go
Comment (toshiiw) [inline k8s.go:70]: I guess we will never use Same. :-)

**Review by toshiiw — COMMENTED** [review]
Not quite finished reviewing. Seems basically okay.

File: ipam.go
Comment (fkakuma) [reply ipam.go:77]: update is the implementation of IpamCache interface type.

File: main.go
Comment (fkakuma) [reply main.go]: Fatal level will call os.Exit(1).
https://godoc.org/github.com/sirupsen/logrus#FatalLevel

File: k8s.go
Comment (fkakuma) [reply k8s.go:575]: update is the implementation of IpamCache interface type.

File: k8s.go
Comment (fkakuma) [reply k8s.go:70]: Maybe. I mainly added it as debugging information.

File: k8s.go
Comment (fkakuma) [reply k8s.go:1]: I will.

File: k8s.go
Comment (fkakuma) [reply k8s.go:172]: I think that it is fixed now and some kind of mechanism may be mad in the future.
At the time the codes should be changed.

File: k8s.go
Comment (fkakuma) [reply k8s.go]: I thought we may use some channels.

**Comment by fkakuma** [comment]
@toshiiw, Thank you for reviewing.

File: k8s.go
Comment (gunjan5) [reply k8s.go]: btw, even with the channels, you don't need a `select` if it's only waiting for one channel. It's a blocking call so `<-time.After...` will block. But I agree with @toshiiw `time.Sleep` is more readable

Comment (fkakuma) [reply k8s.go]: I thought that in the future we may add some channels. But  I will use it as using time.Sleep is simpler in the current codes.

File: k8s.go
Comment (toshiiw) [inline k8s.go]: typo

File: k8s.go
Comment (toshiiw) [inline k8s.go:205]: AFAIK there's no k8s backend specific stuff in this function and we can simply keep using c.server.getNeighborConfigs. Am I missing something?

File: k8s.go
Comment (toshiiw) [inline k8s.go:242]: Unless DeepEqual is much faster than CompareMap, this clause is unnecessary.

File: k8s.go
Comment (toshiiw) [inline k8s.go:267]: You are copying 140 LOC from main.go.

**Review by toshiiw — COMMENTED** [review]
There's some room for improvement in this patch, but we might merge this as is for now. (As a interim fix)

Do you think the efforts are better spend on this client v2 adoption (on calico-bgp-daemon side)?
https://github.com/projectcalico/confd/pull/44

File: k8s.go
Comment (fkakuma) [reply k8s.go]: Oh, Thanks!

File: k8s.go
Comment (fkakuma) [reply k8s.go:205]: bgpconfig is information obtained from k8s backend.

File: k8s.go
Comment (fkakuma) [reply k8s.go:242]: It may be so. But I don't think I will check it.

File: k8s.go
Comment (fkakuma) [reply k8s.go:267]: We may be able to replace them with a common function.
But I made this patch on the assumption that I will refactor later. I also tried not to affect existing code.

File: k8s.go
Comment (fkakuma) [reply k8s.go]: Done.

File: k8s.go
Comment (fkakuma) [reply k8s.go]: Done.

File: k8s.go
Comment (fkakuma) [reply k8s.go:1]: Done.

**Comment by fkakuma** [comment]
I updated codes for some comments.
- Note the file I referred to.
- Use time.Sleep instead of time.After.
- Fix typo.

**Comment by caseydavenport** [comment]
@toshiiw thanks for the review, very helpful :)

I think we should plan to merge this PR and include it in the next Calico v2.x release.

Moving forward, the clientv2 work you found provides a new interface that should allow us to clean up and commonize a lot of this code as well as the etcd based code so they are both using the same logic, but that's a larger rework so I don't want to block this PR on it.

**Comment by caseydavenport** [comment]
@toshiiw ping me here when you think this is good to go and I can press merge. Thanks!

File: main.go
Comment (toshiiw) [reply main.go]: So does "return nil, err".

Comment (fkakuma) [reply main.go]: Ah, it was my misunderstanding. I will fix it.

Comment (fkakuma) [reply main.go]: Done.

**Comment by fkakuma** [comment]
I updated to use return instead of log.Fatal.

**Comment by toshiiw** [comment]
@caseydavenport There are some rooms for improvement but the patch is good enough to be merged. We can clean up later if we wish, or focus on 3.0.

**Review by caseydavenport — APPROVED** [review]
@fkakuma @toshiiw thanks so much for the great work on this.

Let's merge this so we can get it into the next v2.x release, and moving forward we should update this code to be compatible with v3.0.

I think the clientv2 package will help make this code a lot cleaner for the Calico v3.0 version.

**Pull request merged**
